include/fastdds/rtps/security/logging/Logging.h

```diff
@@ -0,0 +1,189 @@
+#define _FASTDDS_RTPS_SECURITY_LOGGING_LOGGING_H_
+
+#include <limits>
+#include <iomanip>
+
+#include "fastdds/rtps/security/logging/LogOptions.h"
+#include "fastdds/rtps/security/logging/BuiltinLoggingType.h"
+#include "fastdds/rtps/security/exceptions/SecurityException.h"
+#include "fastdds/rtps/common/Guid.h"
+
+namespace eprosima {
+namespace fastrtps {
+
+class Publisher;
+
+namespace rtps {
+namespace security {
+
+/**
+ * @brief The LoggerListener class
+ */
+class LoggerListener
+{
+  LoggerListener() = default;
+  ~LoggerListener() = default;
+};
+
+/**
+ * @brief Base class for all security logging plugins.
+ */
+class Logging
+{
+public:
+
+  Logging();
+  virtual ~Logging() = default;
+
+  /**
+   * @brief set_log_options
+   * @param log_options
+   * @return TRUE if successful
+   */
+  bool set_log_options(const LogOptions& log_options, SecurityException& exception);
+
+  /**
+   * @brief get_log_options
+   * @param log_options
+   * @param exception
+   * @return
+   */
+  bool get_log_options(LogOptions& log_options, SecurityException& exception) const;
+
+  /**
+   * @brief enable_logging
+   */
+  bool enable_logging(SecurityException& exception);
+
+  /**
+   * @brief set_listener
+   * @param listener
+   * @return
+   */
+  bool set_listener(LoggerListener* listener, SecurityException& exception);
+
+  /**
+   * @brief log
+   * @param log_level
+   * @param message
+   * @param category
+   */
+  void log(const LoggingLevel log_level,
+           const std::string& message,
+           const std::string& category,
+           SecurityException& exception) const;
+
+  /**
+   * @brief Whether the options are set or not.
+   * @return True if the options are set.
+   */
+  bool options_set() const { return options_set_; }
+
+  /**
+   * @brief Whether the logging is enabled or not.
+   * @return True if the logging is enabled.
+   */
+  bool enabled() const { return logging_enabled_; }
+
+  /**
+   * @brief Return the LoggerListener.
+   * @return A pointer to the (const) LoggerListener.
+   */
+  LoggerListener const* get_listener() const { return listener_; }
+
+  bool set_guid(const GUID_t& guid, SecurityException& exception);
+
+  bool set_domain_id(const uint32_t id, SecurityException& exception);
+
+protected:
+
+  /**
+   * @brief enable_logging_impl
+   * @return
+   */
+  virtual bool enable_logging_impl(SecurityException& /*exception*/) { return true; }
+
+  /**
+   * @brief convert
+   * @param msg
+   * @return
+   */
+  virtual bool convert(const LoggingLevel log_level,
+                       const std::string& message,
+                       const std::string& category,
+                       BuiltinLoggingType& builtin_msg,
+                       SecurityException& exception) const;
+
+  template <typename Stream>
+  bool compose_header(Stream& header,
+                      const BuiltinLoggingType& builtin_msg,
+                      SecurityException& exception) const;
```

Comment:
I assume DDS support will also be streamable? Or do you plan on streaming to an sstream or something? Because that would also work for the file logging. Do we really gain much by having this be a template?